### ISSUE TITLE
filter[ignore_sticky_posts]=false doesn't work with GET /wp/v2/posts

### ISSUE THREAD

**Issue by laustdeleuran** [WP-API/WP-API#2210]
Right now there's is no way to use the `sticky` parameter at all. This feels like a pretty big issue, as this is core functionality in WP.

See threads on [Stack Overflow](http://stackoverflow.com/questions/34970734/wp-api-filter-by-post-schema) and [Wordpress.org](https://wordpress.org/support/topic/wp-api-filter-by-sticky) for examples of how the community seems to want this feature.

Given that `sticky` is already part of the schema, it would make sense if we could just sort by that as a query. 

However, reading #924 and #1801 and #1623, there seems to be discussions around `ignore_sticky_posts`, `respect_sticky_posts` or `include_sticky_posts` **filter** params, but none of these seem to have made it into recent betas.

What's the status on this? And any chance we can get it soon? 

Thanks!

**Comment by danielbachhuber** [comment]
> What's the status on this? And any chance we can get it soon?

You can query by sticky posts with `filter[ignore_sticky_posts]=false`. See #1801

**Comment by laustdeleuran** [comment]
@danielbachhuber As I tried to mention, none of the options discussed in the other issues work.

If I use `filter[ignore_sticky_posts]=false`, [I get sticky and unsticky posts](http://cl.ly/2S0S1j3O3I3E). If I don't use it, [I get sticky and unsticky posts](http://cl.ly/1O3H3K1E3i1n). If I use `filter[ignore_sticky_posts]=true`, [I get sticky and unsticky posts](http://cl.ly/3m0B2M181A15). 

So that parameter seem to make no difference at all in my setup. 

Secondly, even if it worked, I don't really see it covering the basic sticky post use-case, where you want your sticky posts displayed first (or seperately). Imho, the best to get to that point way would be to be able to:
1. Sort by sticky value (`sort_by_stick=ASC` or something) and then date. This might not be the most versatile solution, but would fix the basic use-case in 1 HTTP request.
2. Be able to filter by sticky value (`sticky=true|false` or `filter[sticky]=true|false`) so I can get all sticky posts in one request and all unsticky posts in a seperate request. This seems as the best solution to me.

If I understand `filter[ignore_sticky_posts]` correctly I still have no way of getting a list of the posts that's actually sticky. Right now it seems my only solution is to create a **sticky** category and filter by that, which seems very redundant given that the functionality is already in WP.

I'm using **Version 2.0-beta11**.

**Comment by danielbachhuber** [comment]
> So that parameter seem to make no difference at all in my setup.

When `filter[ignore_sticky_posts]=false`, you'll get the sticky posts first, and then any non-sticky posts afterwards. Can you verify?

Here's our current test case:

```
public function test_get_items_ignore_sticky_posts_by_default() {
    $post_id1 = $this->factory->post->create( array( 'post_status' => 'publish', 'post_date' => '2015-01-01 12:00:00', 'post_date_gmt' => '2015-01-01 12:00:00' ) );
    $post_id2 = $this->factory->post->create( array( 'post_status' => 'publish', 'post_date' => '2015-01-02 12:00:00', 'post_date_gmt' => '2015-01-02 12:00:00' ) );
    $post_id3 = $this->factory->post->create( array( 'post_status' => 'publish', 'post_date' => '2015-01-03 12:00:00', 'post_date_gmt' => '2015-01-03 12:00:00' ) );
    stick_post( $post_id2 );
    // No stickies by default
    $request = new WP_REST_Request( 'GET', '/wp/v2/posts' );
    $response = $this->server->dispatch( $request );
    $data = $response->get_data();
    $this->assertEquals( array( $this->post_id, $post_id3, $post_id2, $post_id1 ), wp_list_pluck( $data, 'id' ) );
    // Permit stickies
    $request = new WP_REST_Request( 'GET', '/wp/v2/posts' );
    $request->set_param( 'filter', array( 'ignore_sticky_posts' => false ) );
    $response = $this->server->dispatch( $request );
    $data = $response->get_data();
    $this->assertEquals( array( $post_id2, $this->post_id, $post_id3, $post_id1 ), wp_list_pluck( $data, 'id' ) );
}
```

> Be able to filter by sticky value (sticky=true|false or filter[sticky]=true|false) so I can get all sticky posts in one request and all unsticky posts in a seperate request. This seems as the best solution to me.

This isn't currently supported by core, and won't be directly supported by the REST API. However, you could write a custom endpoint to get the ids of sticky posts, and then use the `include` param to just get those sticky posts.

**Comment by laustdeleuran** [comment]
The `filter[ignore_sticky_posts]` parameter has no effect on the output of my `/wp-json/wp/v2/posts/` endpoint. The posts come out in the exact same order and no matter if they have the `sticky` flag or not.

Will try to deactivate all my other plugins to make sure nothing is interfering. 

---

I figured I could potentially write a custom endpoint for it, but I find the documentation a bit hard to digest at the moment (as would be expected by a project in development). If all else fails, I'll try that.

**Comment by laustdeleuran** [comment]
I deactivated all my other plugins and switched back to the _twentysixteen_ theme and [the parameter still has no impact on the output](http://cl.ly/0s2g1R073t22).

**Comment by danielbachhuber** [comment]
Ok. I was able to reproduce this, and tracked down the problem. Unfortunately, it's not going to be an easy fix. For now, your best bet would be to expose sticky post ids through an endpoint, use `include` to fetch just those posts, and then use a second query with `exclude` to get remaining posts minus the sticky posts.

Ultimately, `filter[ignore_sticky_posts]=false` is ignored because sticky posts only kick in when `WP_Query->is_home=true` ([ref](https://core.trac.wordpress.org/browser/tags/4.4.1/src/wp-includes/query.php#L3724)). However, `is_home=false` when `REST_REQUEST` is defined ([ref](https://core.trac.wordpress.org/browser/tags/4.4.1/src/wp-includes/query.php#L1814)). On top of all of this, our tests pass because `is_home=true` when our tests run.

**Comment by laustdeleuran** [comment]
Okay, thanks for looking into this! 

Can you point me in any direction as to how to create my own endpoint? As I'm not a "real" PHP dev, I find [the documentation](http://v2.wp-api.org/extending/adding/) a bit difficult to understand.

Thanks again!

**Comment by laustdeleuran** [comment]
> Can you point me in any direction as to how to create my own endpoint? As I'm not a "real" PHP dev, I find the documentation a bit difficult to understand.

I figured this out, so no worries. Basically, this is what I'm doing:

``` php
function get_sticky_posts() {
    $posts = get_posts(
        array(
            'post__in' => get_option('sticky_posts')
        )
    );

    if (empty($posts)) {
        return null;
    }

    return $posts;
}
add_action( 'rest_api_init', function () {
    register_rest_route( 'THEME_NAME/v1', '/sticky', array(
        'methods' => 'GET',
        'callback' => 'get_sticky_posts',
    ));
});
```

Thanks for looking into this and pointing me in the right direction to get a temporary fix for my problem now.

**Comment by danielbachhuber** [comment]
Fixing this will require a core patch https://core.trac.wordpress.org/ticket/35907

**Comment by aaronjorbin** [comment]
@danielbachhuber core patch is in - https://core.trac.wordpress.org/changeset/36696

**Comment by danielbachhuber** [comment]
Moving this back into the 2.0 bucket, until there's an implementation in core.

**Comment by dleeward** [comment]
I'm trying to set sticky to true for a custom post type in a POST but it doesn't work and doesn't show in the context=edit response. Does the API support setting sticky for CPTs?

**Comment by kadamwhite** [comment]
bug triage slack discussion: https://wordpress.slack.com/archives/core-restapi/p1473951755001631

**Comment by krogsgard** [comment]
This is causing a discussion of sticky posts more generally and how they should be treated by the API. Looks like @joehoyle is making a separate, now unrelated ticket, to be able to flag `sticky=true` to get _just_ sticky posts, and there are additional comments in the discussion linked above about how we should handle sticky posts in regular queries.

**Comment by kadamwhite** [comment]
Closing out as resolved by #2708 / superseded by #2764
